### ISSUE TITLE
Wrong signed AuthnRequest with redirect binding

### ISSUE THREAD

**Issue by Budovi** [IdentityPython/pysaml2#579]
The `Saml2Client.prepare_for_negotiated_authenticate` method generates wrong AuthnRequest when the binding *for the request* is `BINDING_HTTP_REDIRECT` and the request is supposed to be signed.

Issue mentioned also #163 , but was unable to find any references of how this is supposed to be handled in the current code path. Tried to botch a workaround instead, but the library is in so much technical debt that I'm unable to do so.

Other related issues include #334 possibly #570 .

## Code Version
release 4.6.5

## Expected Behavior
The request queries contain initialized `SAMLRequest`, `Signature` and `SigAlg` parameters and optional `RelayState`.

## Current Behavior
Request consist of `SAMLRequest` that is deflated and encoded with the signature *embedded* in the xml. Optional `RelayState` works fine.

## Possible Solution
Maybe this is handled correctly in the provided examples, but those duplicate and / or substitute the functionality that is already present (or should be) in the library IMHO. If this is the case, the `Saml2Client` functionality should be fixed.

I cannot use the provided examples in my project since it is not a typical WSGI application (and honestly, looking at it, I don't want either). So I didn't test either.

## Steps to Reproduce
1. Configure a SP with `"authn_requests_signed": True`. In my case, the SP uses remote metadata obtained from a federation, but I don't think this has anything to do with the issue.
2. Initialize the client with `saml_sp = Saml2Client(config_file="<config_file_path>")`.
3. Through a WAYF discovery, obtain a target `entityID`. This IdP shall support SSO using the redirect binding.
4. Call `saml_sp.prepare_for_negotiated_authenticate(entityid=entityID, relay_state=RelayState)`.

I'm sorry for not being more helpful, but I currently have no more time to invest in this project, have to get the code to work ASAP.

--------
Haven't had luck with enforcing the POST binding either, the IdP refuses to accept the generated request. I currently have no idea why since I'm not in control of the IdP and contacting support takes time. The generated request seems to be a valid AuthnRequest. So it may not be an issue with the pysaml2, just pointing it out if you spot anything weird.

**Comment by c00kiemon5ter** [comment]
Hello, this has up before and is one of the things that need to be refactored to make this easier to handle. In your case, I think you need the following things:
- set `sign=True` when calling `prepare_for_negotiated_authenticate`, and
- pass a `sigalg` as a `kwargs` entry
- you can also use `prepare_for_authenticate` if you do not care about the chosen binding

```python
import saml2.xmldsig as ds


saml_sp.prepare_for_authenticate(
  entityid=entityID,
  relay_state=RelayState,
  sign=True,
  kwargs={"sigalg": ds.SIG_RSA_SHA1},
)
```

**Comment by peppelinux** [comment]
This could be considered for a better understanding of the complexity of this issue.
https://github.com/IdentityPython/pysaml2/issues/713#issuecomment-687456116

If the signature is in HTTP-REDIRECT url args it shouldn't also be in the saml2 payload (xml).
Another test to do would be using the followin asset:

1. A not-pysaml2 SP (eg. Shibboleth SP)
2. A pysaml2 IdP with WantRequestSigned and HTTP-REDIRECT

AFAIK ShibSP put Signature, correctly, i url args and not these in saml2 payload. Pysaml2 IdP instead only checks for those that would have been (incorrect!) in the saml2 payload.

Still have to re-test this for a better acknowledgement.
Need help if possibile

**Comment by peppelinux** [comment]
Moved here:
https://github.com/IdentityPython/pysaml2/issues/633

Thank you to all for acknowledgments

**Comment by c00kiemon5ter** [comment]
This should be fixed by 2f756ba
